core/kernel/transfer_list.c

```diff
@@ -0,0 +1,535 @@
+
+	if (!tl)
+		return NULL;
+
+	max_tl_ev = (vaddr_t)tl + tl->max_size;
+	tl_ev = (vaddr_t)tl + tl->size;
+	ev = tl_ev;
+
+	/*
+	 * skip the step 1 (optional step)
+	 * new TE will be added into the tail
+	 */
+	if (ADD_OVERFLOW(sizeof(*te), data_size, &sz) ||
+	    ADD_OVERFLOW(ev, sz, &ev) ||
+	    ROUNDUP_OVERFLOW(ev, TRANSFER_LIST_GRANULE, &ev) ||
+	    ev > max_tl_ev) {
+		return NULL;
+	}
+
+	te = (struct transfer_list_entry *)tl_ev;
+	te->tag_id = tag_id;
+	te->reserved0 = 0;
+	te->hdr_size = sizeof(*te);
+	te->data_size = data_size;
+	tl->size += ev - tl_ev;
+
+	if (data) {
+		/* get TE data pointer */
+		te_data = transfer_list_entry_data(te);
+		if (!te_data)
+			return NULL;
+
+		memmove(te_data, data, data_size);
+	}
+
+	transfer_list_update_checksum(tl);
+
+	return te;
+}
+
+/*******************************************************************************
+ * Add a new transfer entry into a transfer list with specified new data
+ * alignment requirement
+ * Compliant to 2.4.4 of Firmware handoff specification (v0.9)
+ * 'argument' is specified as a power of two according to the spec
+ * Return pointer to the added transfer entry or NULL on error
+ ******************************************************************************/
+struct transfer_list_entry *
+transfer_list_add_with_align(struct transfer_list_header *tl, uint16_t tag_id,
+			     uint32_t data_size, const void *data,
+			     uint8_t alignment)
+{
+	struct transfer_list_entry *te = NULL;
+	vaddr_t tl_ev = 0;
+	vaddr_t ev = 0;
+	vaddr_t new_tl_ev = 0;
+	size_t dummy_te_data_sz = 0;
+
+	if (!tl)
+		return NULL;
+
+	tl_ev = (vaddr_t)tl + tl->size;
+	ev = tl_ev + sizeof(struct transfer_list_entry);
+
+	if (!IS_ALIGNED(ev, TL_ALIGNMENT_FROM_ORDER(alignment))) {
+		/*
+		 * TE data address is not aligned to the new alignment
+		 * fill the gap with an empty TE as a placeholder before
+		 * adding the desire TE
+		 */
+		new_tl_ev = ROUNDUP(ev, TL_ALIGNMENT_FROM_ORDER(alignment)) -
+			    sizeof(struct transfer_list_entry);
+		dummy_te_data_sz =
+			new_tl_ev - tl_ev - sizeof(struct transfer_list_entry);
+		if (!transfer_list_add(tl, TL_TAG_EMPTY, dummy_te_data_sz,
+				       NULL)) {
+			return NULL;
+		}
+	}
+
+	te = transfer_list_add(tl, tag_id, data_size, data);
+
+	if (alignment > tl->alignment) {
+		tl->alignment = alignment;
+		transfer_list_update_checksum(tl);
+	}
+
+	return te;
+}
+
+/*******************************************************************************
+ * Search for an existing transfer entry with the specified tag id from a
+ * transfer list
+ * Return pointer to the found transfer entry or NULL on error
+ ******************************************************************************/
+struct transfer_list_entry *transfer_list_find(struct transfer_list_header *tl,
+					       uint16_t tag_id)
+{
+	struct transfer_list_entry *te = NULL;
+
+	do {
+		te = transfer_list_next(tl, te);
+	} while (te && (te->tag_id != tag_id || te->reserved0));
```

Comment:
hmmm, this loop will stop when `te->tag_id == tag_id && !te->reserved0`, otherwise `NULL` when not found, am I right?

Reply:
If it can't be anything but 0 there's no point in checking if it really is. My point is that if for some reason it isn't 0 we'll return a `te` with an unexpected tag.

Reply:
OK. Sounds good. I can ignore checking `reserved0` here.